### PR TITLE
feat: Adopt filter types from mongodb and enhance strictness

### LINKED ISSUE #410: Own dot-notation filter types
The MongoDB team is planning to remove the default dot-notation support from the `Filter` type in the upcoming v5 of the driver.

See: https://github.com/mongodb/node-mongodb-native/pull/3520

We want to adopt those types inside `Papr`, in order to keep Papr filters type safe.

### PR DESCRIPTION
**Pull request by avaly**
Closes #410 #407

This PR brings the strict filter types from `mongodb` into our repository, which allows us to remove the `Document` type and enhance the strictness of these types.

Known divergence from the `mongodb` types:

- added `$expr` support to the root filter operators
- `$elemMatch` is applied only for array fields
- removed support for recursive types in the dot-notation path algorithm
- reduced maximum depth level to 6 in the dot-notation path algorithm
- added support for non-indexed array elements in the dot-notation path algorithm, e.g. `foo.bar` (where `foo` is an array of objects)
- added support for indexed array elements in the dot-notation path algorithm, e.g. `foo.0` (where `foo` is an array of objects)

### CODE REVIEW

**Review by ejmartin504 — COMMENTED** [review]
I can't really say that I fully understand all the TS, but a quick look doesn't raise any red flags for me.

You mentioned a few things in the PR description that I didn't quite follow, would you mind expanding on these, or pointing where they are in the code, for those following along at home?

- _added $expr support to the root filter operators_ --  What is this? Why was this not supported by mongo?
- _added support for non-indexed array elements in the dot-notation path algorithm, e.g. -foo.bar (where foo is an array of objects)_ -- What is the use case for this? Is this query pattern supported by the mongo server?
- _added support for indexed array elements in the dot-notation path algorithm, e.g. foo.0 (where foo is an array of objects)_ -- surprised this wasn't supported by mongo if this is a valid query pattern, is this query pattern supported by the mongo server? 

Also, I'm curious how we plan to keep these types in sync with mongodb types themselves.

**Comment by avaly** [comment]
> _added $expr support to the root filter operators_ --  What is this? Why was this not supported by mongo?

https://www.mongodb.com/docs/v5.3/reference/operator/query/expr/

`$expr` is a root filter operator. It was never defined in the `mongodb` types.

This was possible to use because of the `Document` being used everywhere in those types, which accepted any unknown property.

> _added support for non-indexed array elements in the dot-notation path algorithm, e.g. -foo.bar (where foo is an array of objects)_ -- What is the use case for this? Is this query pattern supported by the mongo server?

https://www.mongodb.com/docs/v5.3/tutorial/query-array-of-documents/#specify-a-query-condition-on-a-field-embedded-in-an-array-of-documents

> _added support for indexed array elements in the dot-notation path algorithm, e.g. foo.0 (where foo is an array of objects)_ -- surprised this wasn't supported by mongo if this is a valid query pattern, is this query pattern supported by the mongo server?

https://www.mongodb.com/docs/v5.3/tutorial/query-arrays/#query-for-an-element-by-the-array-index-position

File: docs/recipes.md [hunk 1]
```diff
@@ -199,15 +199,16 @@ Starting with `mongodb` v4.3.0, these types were enhanced to support dot notatio
 
 However, `mongodb` v5.0.0 [removed these types](https://github.com/mongodb/node-mongodb-native/blob/main/etc/notes/CHANGES_5.0.0.md#dot-notation-typescript-support-removed-by-default) as the default ones used in their methods and reverted to the old ones without dot notation support. The previous enhanced types were not removed, instead they were renamed to `StrictFilter` and `StrictUpdateFilter`, but they aren't referenced in any of their methods.
 
-Papr is using the strict types to provide type safety for all query and update filters.
+Papr v11 has adopted and enhanced these strict types to provide type safety for all query and update filters.
 
-This comes with a caveat: whenever you need to interact with the `mongodb` driver collections, you need to cast filter types to their simple counterparts, since `Filter` is not compatible with `StrictFilter`.
+This comes with a caveat: whenever you need to interact with the `mongodb` driver collections, you need to cast filter types to their simple counterparts, since `Filter` is not compatible with `PaprFilter`.
```
Comment (marafat) [inline docs/recipes.md:204]: I assume as far as the compiled JS is concerned, we will never be making a query with an "invalid syntax" as a result of this forced casting? right?

File: src/mongodbTypes.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
+import type {
```
Comment (marafat) [inline src/mongodbTypes.ts:1]: if re-ordering the functions in this file is not preferred, does it make sense to disable that lint rule on this entire file?

File: src/mongodbTypes.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  >;
+} & {
+  [Property in `${NestedPathsOfType<TSchema, Record<string, any>[]>}.$${
+    | ''
+    | `[${string}]`}.${string}`]?: any;
+} & {
+  [Property in Join<NestedPaths<TSchema, []>, '.'>]?: PropertyType<TSchema, Property>;
+};
+
+export interface PaprUpdateFilter<TSchema> {
```
Comment (marafat) [inline src/mongodbTypes.ts:150]: I can't help but thinking of extending the mongo type instead of defining a standalone one? What do you think? we can `Omit` or union with whatever we need but I feel, IMO, it would be more descriptive in code of how we diverge from the mongo types...

File: src/utils.ts [hunk 4]
```diff
@@ -1,19 +1,9 @@
 import { ObjectId } from 'mongodb';
```
Comment (marafat) [inline src/utils.ts:1]: Those change are hard to follow, I wonder if we can add some basic tests using `ts-expect`?

**Review by marafat — APPROVED** [review]

File: src/mongodbTypes.ts (hunk 3 above)
Comment (avaly) [reply src/mongodbTypes.ts:150]: We can't TBH, at least it would be extremely messy, since their usage of `Document` is embedded in a lot of nested types used here.

File: src/utils.ts (hunk 4 above)
Comment (avaly) [reply src/utils.ts:1]: Which changes exactly? All of these new types added here are using inside `mongodbTypes`, which has extensive type tests. But I can write some more tests for these helpers.

File: src/mongodbTypes.ts (hunk 3 above)
Comment (avaly) [reply src/mongodbTypes.ts:150]: For example, this is the `StrictUpdateFilter` definition in mongodb: https://github.com/mongodb/node-mongodb-native/blob/main/src/mongo_types.ts#L518-L542

We can use it like `Omit<StrictUpdateFilter<TSchema>, Document>`, because `Document` is defined as https://github.com/mongodb/js-bson/blob/6ef100083328b2eeeb3c6a536985f494f919071f/src/bson.ts#L58-L61

Comment (avaly) [reply src/mongodbTypes.ts:152]: FWIW, I did that already for `PaprBulkWriteOperation` because it was possible there, since those types weren't merging `Document`.

File: src/utils.ts (hunk 4 above)
Comment (avaly) [reply src/utils.ts:3]: I've added tests for the new types I added in this file.
